### PR TITLE
Reduce noise from LMS automated media device mounting script, service

### LINKED ISSUE #733: Fix broken LMS client connection, or stop spamming logs
It's not clear what's prompting [this to happen](https://github.com/micro-nova/AmpliPi/pull/682#issuecomment-2135447077), but it ought not. Either correct the invocation involved here, or gracefully handle the disconnect case instead of spamming the logs with tracebacks.

Related to #682.

### PR DESCRIPTION
**Pull request by SteveMicroNova**
### What does this change intend to accomplish?
Closes #733 by making fewer logs and only running the noisy part when it's meant to

### Checklist

* [x] Have you tested your changes and ensured they work?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] If applicable, have you updated the CHANGELOG?

### CODE REVIEW

**Comment by SteveMicroNova** [comment]
Needs to be tested with an actual USB drive during LMS mode
I do not currently have a USB drive

**Comment by rtertiaer** [comment]
might make sense to roll #766 into this as well

**Comment by rtertiaer** [comment]
I'll review this when you've tested it

**Comment by SteveMicroNova** [comment]
> Needs to be tested with an actual USB drive during LMS mode
> I do not currently have a USB drive

I recovered my USB drive at end of day yesterday, after some further tweaks it now works just fine, even caught some other noisy errors that were nonissues caused by a race condition during startup (will point those out in a self review shortly)

File: scripts/edit_media_directories.py [hunk 1]
```diff
@@ -20,25 +31,42 @@ def get_usb_drives(logger: logging.Logger):
...
+        for drive in usb_drives:
+          drives += f"{drive}, "
+        logger.info(f"Detected drives: {drives}")
+        logger.info("If you wish to mount these to the internal LMS server, please go to Config and set LMS Mode")
+  except urllib3.exceptions.MaxRetryError:
+    # This error is extremely common during startup, but isn't actually broken
+    # It seemingly only occurs on startup, leading me to believe it is a race condition with other things starting up
+    logger.error("LMS Automated Drive Mounting has encountered an error due to a race condition during startup, if this persists (or happened outside of startup) please contact support@micro-nova.com with a copy of these logs")
+  except Exception as e:
+    logger.error(f"LMS Automated Drive Mounting has encountered an error: {e}")
```
Comment (SteveMicroNova) [inline scripts/edit_media_directories.py:68]: Note the comment here, the requests.post above will throw errors if it attempts to run while LMS is starting, but will be fine once LMS is finished loading

File: scripts/edit_media_directories.py [hunk 2]
```diff
@@ -5,11 +5,22 @@
 import requests
 import logging
 import sys
+import subprocess
+import urllib3
+import urllib3.exceptions
+
+def CheckLMSMode():
```
Comment (linknum23) [inline scripts/edit_media_directories.py:12]: CamelCase is typically used only for classes in python. Is there a specific reason this and the LMSMode variable below are in camel case?

File: scripts/edit_media_directories.py [hunk 3]
```diff
@@ -20,25 +31,43 @@ def get_usb_drives(logger: logging.Logger):
...
-  for d, drive in enumerate(usb_drives):
-    data[f"pref_mediadirs{d}"] = drive
-    data[f"pref_ignoreInAudioScan{d}"] = 1
-
-  logger.info("Adding drives to LMS settings...")
-  requests.post(url, data=data, verify=False, timeout=5)
-  logger.info("Drives added, LMS scanning now.")
+  try:
+    usb_drives = get_usb_drives(logger) # Get list of available drives
+    LMSMode = CheckLMSMode()
```
Comment (linknum23) [inline scripts/edit_media_directories.py:36]: This assignment reads confusing it looks like you are assigning a class/type from an instantiated class.

File: scripts/edit_media_directories.py (hunk 2 above)
Comment (SteveMicroNova) [reply scripts/edit_media_directories.py:12]: CamelCase is simply how I write code, back when I was taught in school originally that's how we did everything and that practice has stuck with me

File: scripts/edit_media_directories.py [hunk 4]
```diff
@@ -20,25 +31,43 @@ def get_usb_drives(logger: logging.Logger):
...
+      logger.info("Drives added, LMS scanning now.")
+    else:
+      if len(usb_drives) > 0:
+        drives = ""
+        for drive in usb_drives:
+          drives += f"{drive}, "
+        logger.info(f"Detected drives: {drives}")
+        logger.info("If you wish to mount these to the internal LMS server, please go to Config and set LMS Mode")
+  except (urllib3.exceptions.MaxRetryError, urllib3.exceptions.NewConnectionError, requests.exceptions.ConnectionError):
+    # This error seemingly only occurs on startup, leading me to believe it is a race condition
```
Comment (linknum23) [inline scripts/edit_media_directories.py:64]: Have we made any attempt to delay this service? Maybe make it dependent on the Logitech media server?

Comment (SteveMicroNova) [reply scripts/edit_media_directories.py]: I've made a few attempts in my last few commits, between creating the function that checked if lms was running to the systemd dependency attempts, it gives the same results of having this error print out 4 times during startup:
>Jun 25 15:30:10 lamplipi udisks2-listener.sh[551]: LMS Automated Drive Mounting has encountered an error: HTTPConnectionPool(host='localhost', port=9000): Max retries exceeded with url: /settings/server/basic.html (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x75db4390>: Failed to establish a new connection: [Errno 111] Connection refused'))

If there's another way to do this that I'm not aware of, please let me know

**Comment by linknum23** [comment]
I think the best way would be to add a dependency in the service file. I
think you can use the after keyword:
Here's an example that waits for the  network to be up and ssh-keygen to be
running

    After=network.target sshd-keygen.service

I assume you could do the same thing for logitechmediaserver.service

You may also want to check out using the Wants and Requires keywords as
well to specify the LMS dependency.

**Comment by SteveMicroNova** [comment]
I think that https://github.com/micro-nova/AmpliPi/pull/768/commits/8fc3525765ec69d12c063f06d13499afbb06b043 should solve the disjointed startup times, will test when I have access to a unit to test on tomorrow

**Comment by SteveMicroNova** [comment]
I've attempted many permutations of "Only start when logitechmediaserver.service is running" between python code and systemd scripting but none have kept the script from winning the race condition and printing exactly 4 copies of the same error into the logs, it seems that the logitechmediaserver.service advertises itself as good to go a few seconds before that's true

Unless there's some other way to hook it in, I'll suggest to either:
1. live with exactly 4 error messages on every boot cycle
2. suppress the errors by catching without reraising
3. have a time.sleep(8) at the start of the script so that LMS fully loads rather than getting 4 of the same error with 2 seconds between each one

**Comment by SteveMicroNova** [comment]
I contacted Ryan separately about my concerns and he sent me this: [https://www.slingacademy.com/article/python-requests-auto-retry-for-failed-http-requests/](https://www.slingacademy.com/article/python-requests-auto-retry-for-failed-http-requests/)

Which solved the issue entirely. No more random LMS errors in the console.

**Review by rtertiaer — APPROVED** [review]
:sunglasses:

**Pull request merged**